keps/sig-network/20180612-ipv4-ipv6-dual-stack.md

```diff
@@ -401,13 +401,13 @@ A new [kube-proxy configuration](https://kubernetes.io/docs/reference/command-li
 ```
 Only the first CIDR for each IP family will be used; all others will be ignored.
 
-### CoreDNS Operation
+### DNS Operation
 
-CoreDNS will need to make changes in order to support the plural form of endpoint addresses. Some other considerations of CoreDNS support for dual-stack:
+DNS will need to support the plural form of endpoint addresses. Some other considerations of DNS support for dual-stack:
 
-- Because service IPs will remain single-family, pods will continue to access the CoreDNS server via a single service IP. In other words, the nameserver entries in a pod's /etc/resolv.conf will typically be a single IPv4 or single IPv6 address, depending upon the IP family of the cluster's service CIDR.
-- Non-headless Kubernetes services: CoreDNS will resolve these services to either an IPv4 entry (A record) or an IPv6 entry (AAAA record), depending upon the IP family of the cluster's service CIDR.
-- Headless Kubernetes services: CoreDNS will resolve these services to either an IPv4 entry (A record), an IPv6 entry (AAAA record), or both, depending on the service's `ipFamily`.
+- Because service IPs will remain single-family, pods will continue to access the DNS server via a single service IP. In other words, the nameserver entries in a pod's /etc/resolv.conf will typically be a single IPv4 or single IPv6 address, depending upon the IP family of the cluster's service CIDR.
+- Non-headless Kubernetes services: DNS will resolve these services to either an IPv4 entry (A record) or an IPv6 entry (AAAA record), depending upon the IP family of the cluster's service CIDR.
+- Headless Kubernetes services: DNS will resolve these services to either an IPv4 entry (A record) or an IPv6 entry (AAAA record), depending on the service's `ipFamily`, defaulting to IPv4 if `ipFamily` is not specified.
```

Comment:
I think that we have to preserve backwards compatible behavior of the endpoints object (where you can have both types of addresses), just as general Kubernetes API guarantee.

That implies that anything we do with ipFamily has to allow users to preserve the existing behavior, and thus downstream components have to tolerate mixed addresses and offer meaningful extension.

I would prefer that DNS be changed to use ipFamily as a filter, and DNS treat nil ipFamily as "allow all addresses".

Reply:
Just to clarify, does this discussion only apply to headless and external name services? Can we safely state that all Services handled by kube-proxy will be single family since they will all have a single ClusterIP and/or IPFamily? Is there any precedent for supporting multi-family Endpoints behind single family Services before the dual stack work? If we can limit this to Services ignored by kube-proxy, I think it would significantly limit the scope and complexity of the needed changes. I do agree that for the sake of backwards compatibility, the lack of an IPFamily should be treated by DNS as "allow all addresses."

Reply:
seems we are on the same page, will reword accordingly

Reply:
@smarterclayton I just realized that there is a different behavior on headless services depending on using selectors or not. This is the bug that Dan found, currently headless service with selectors filters based on the ipFamily and if nil, it defaults to IPv4.
This is due to the function that returns the endpoints based on the pods, only returns one IP per pod. If we want to support this for headless return multiple endpoints per pod we have to modify this function too.
Do we want to modify the behavior of `headless service with selectors` to support `ipFanily=nil`?

ref: https://github.com/kubernetes/kubernetes/pull/87608

Reply:
> Do we want to modify the behavior of headless service with selectors to support ipFamily=nil?

I think that might be appropriate, if DNS is expected to report A and AAAA when ipfamily is nil.

Reply:
Yep, I'd agree with that since kube-dns is already responding with both A and AAAA records in some cases (https://github.com/kubernetes/dns/blob/master/vendor/github.com/skynetservices/skydns/server/server.go#L476-L479).

Reply:
ok, let me try to send a PR with the modifications needed so we can observe the implications

Reply:
I've updated this PR with the last @thockin suggestion of keep the current endpoints controller defaulting to IPv4 and modifying the endpointslice controller to return both A and AAAA